git/odb/blob.go

```diff
@@ -68,3 +68,18 @@ func (b *Blob) Close() error {
 	}
 	return b.closeFn()
 }
+
+// Equal returns whether the receiving and given blobs are equal, or in other
+// words, whether they are represented by the same SHA-1 when saved to the
+// object database.
+func (b *Blob) Equal(other *Blob) bool {
+	if (b == nil) != (other == nil) {
+		return false
+	}
+
+	if b != nil {
+		return b.Contents == other.Contents &&
```

Comment:
I'm not sure comparing `io.Reader` is the right solution here. I'd expect a test like this to pass:

```go
func TestBlobEqualReturnsTrueWithUnchangedButReadContents(t *testing.T) {
	c1 := strings.NewReader("Hello, world!")
	c2 := strings.NewReader("Hello, world!")

	b1 := &Blob{Size: int64(c1.Len()), Contents: c1}
	b2 := &Blob{Size: int64(c2.Len()), Contents: c2}

	assert.True(t, b1.Equal(b2))
}
```

Based on the way that the blob rewrite function usually returns a new `*Blob`, I think this will always fail the `Equal()` check, unless both `Contents` are pointers to the same reader.

https://github.com/git-lfs/git-lfs/blob/7aeeec245c69984279fe2531b077e41e2d00ad74/commands/command_migrate_import.go#L31-L49

I suppose this works out in practice, since any call to `BlobFn` means that blob is definitely being converted.

Reply:
Great question -- you're absolutely right. Two `io.Reader`'s that yield the same content but are instantiated differently will fail the equality check.

That's fine in practice here, since as you pointed out, all calls to the `BlobFn` mean the blob is definitely being converted. That said, there are cases in the 'info' migrator where the contents of the blob does not change:

https://github.com/git-lfs/git-lfs/blob/cdd0cb70a45db94a93741013e7b351c003784be5/commands/command_migrate_info.go#L70-L91

A more complete implementation might involve some buffering elsewhere, and keeping track of the SHA as it changes, but I think that this will do for now.